File: e2e/src/ez_cdc_e2e/tui/banner.py
"""Banner rendering for the EZ-CDC CLI.

Two banner variants:

  - Banner A (ANSI Shadow): large block-style ASCII art with a vertical
    gradient. Used in the main menu (`ez-cdc` with no subcommand) and in
    `ez-cdc quickstart`. Needs at least 56 columns; falls back to banner D
    when the terminal is narrower.

  - Banner D (gradient bar): compact two-line gradient bar with centered
    text. Used in all other subcommands (`verify`, `load`, `up`, `down`,
    `logs`, `status`). Fits in any terminal >= 40 columns.

Gradients are interpolated between primary.600 (#2563EB) and primary.400
(#60A5FA), matching the brand logo gradient.
"""

from __future__ import annotations

import shutil

from rich.console import Group, RenderableType
from rich.text import Text

from .theme import PRIMARY_400, PRIMARY_600, GRAY_400, interpolate_hex


# ── Banner A: ANSI Shadow block art ──────────────────────────────────────────

# 6 lines, 44 columns wide. "EZ-CDC" rendered in ANSI Shadow figlet font.
_BANNER_A_LINES = [
    "███████╗███████╗     ██████╗██████╗  ██████╗",
    "██╔════╝╚══███╔╝    ██╔════╝██╔══██╗██╔════╝",
    "█████╗    ███╔╝     ██║     ██║  ██║██║     ",
    "██╔══╝   ███╔╝      ██║     ██║  ██║██║     ",
    "███████╗███████╗    ╚██████╗██████╔╝╚██████╗",
    "╚══════╝╚══════╝     ╚═════╝╚═════╝  ╚═════╝",
]

_BANNER_A_MIN_WIDTH = 56  # enough padding to look balanced
_BANNER_A_TAGLINE = "Real-time CDC, radically simplified"


def render_banner_a() -> RenderableType:
    """Render the ANSI Shadow banner with a vertical gradient.

    Returns a rich Renderable (Group of Text lines). Each line of the ASCII
    art gets its own gradient color, from primary.600 at the top to
    primary.400 at the bottom.

    If the terminal is narrower than 56 columns, falls back to banner D.
    """
    width = shutil.get_terminal_size((80, 24)).columns
    if width < _BANNER_A_MIN_WIDTH:
        return render_banner_d()

    colors = interpolate_hex(PRIMARY_600, PRIMARY_400, len(_BANNER_A_LINES))

    lines: list[RenderableType] = [Text("")]  # top padding
    for line, color in zip(_BANNER_A_LINES, colors):
        t = Text(line, style=f"bold {color}")
        t.pad_left(4)
        lines.append(t)

    lines.append(Text(""))  # space before tagline
    tagline = Text(_BANNER_A_TAGLINE, style=GRAY_400, justify="center")
    # Center tagline under the ASCII art (the art is ~44 cols, padded left 4,
    # so centering around col 4 + 22 = 26). Use pad_left calc instead of
    # justify="center" so we center *under the art*, not the whole terminal.
    tagline_padding = 4 + (44 - len(_BANNER_A_TAGLINE)) // 2
    if tagline_padding > 0:
        tagline = Text(" " * tagline_padding + _BANNER_A_TAGLINE, style=GRAY_400)
    lines.append(tagline)

    lines.append(Text(""))  # bottom padding
    return Group(*lines)


# ── Banner D: gradient bar ───────────────────────────────────────────────────

_BANNER_D_BAR_WIDTH = 52
_BANNER_D_TEXT = "EZ-CDC  •  Real-time CDC, radically simplified"
_BANNER_D_PAD_LEFT = 4


def _make_gradient_bar(char: str, width: int) -> Text:
    """Build a horizontal gradient bar of `width` chars using the given char.

    Each character is styled with an interpolated color between primary.600
    and primary.400, producing a smooth left-to-right gradient.
    """
    colors = interpolate_hex(PRIMARY_600, PRIMARY_400, width)
    t = Text()
    t.append(" " * _BANNER_D_PAD_LEFT)
    for color in colors:
        t.append(char, style=color)
    return t


def render_banner_d() -> RenderableType:
    """Render the compact gradient bar banner.

    Layout:
        (blank)
        ▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
                EZ-CDC  •  Real-time CDC, radically simplified
        ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
        (blank)

    The bars use ▄ (upper half block) above the text and ▀ (lower half block)
    below, creating the visual effect of a solid bar hugging the text.
    """
    width = shutil.get_terminal_size((80, 24)).columns
    bar_width = min(_BANNER_D_BAR_WIDTH, max(40, width - _BANNER_D_PAD_LEFT - 2))

    top_bar = _make_gradient_bar("▄", bar_width)
    bot_bar = _make_gradient_bar("▀", bar_width)

    # Build text line: "EZ-CDC" in brand + "•" in dim + tagline in tagline
    text_line = Text()
    text_padding = _BANNER_D_PAD_LEFT + max(0, (bar_width - len(_BANNER_D_TEXT)) // 2)
    text_line.append(" " * text_padding)
    text_line.append("EZ-CDC", style=f"bold {PRIMARY_400}")
    text_line.append("  •  ", style=GRAY_400)
    text_line.append("Real-time CDC, radically simplified", style=GRAY_400)

    return Group(
        Text(""),
        top_bar,
        Text(""),
        text_line,
        Text(""),
        bot_bar,
        Text(""),
    )

File: e2e-cli/src/tui/banner.rs
//! Banner rendering for the EZ-CDC CLI.
//!
//! Two variants:
//! - `render_banner()` — ANSI Shadow block art with vertical gradient. Used
//!   in the main menu and quickstart.
//! - `render_banner_compact()` — single-line styled "ez-cdc" + version. Used
//!   in subcommands like verify, up, down, etc.
//!
//! Gradients interpolated between primary.600 and primary.400.

use console::style;

use super::theme::{interpolate_hex, GRAY_400, PRIMARY_400, PRIMARY_600};

// ── ANSI Shadow block art ─────────────────────────────────────────────────

// 6 lines, 44 columns wide. "EZ-CDC" rendered in ANSI Shadow figlet font.
const BANNER_LINES: &[&str] = &[
    "\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}     \u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557} \u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}",
    "\u{2588}\u{2588}\u{2554}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}\u{255A}\u{2550}\u{2550}\u{2588}\u{2588}\u{2588}\u{2554}\u{255D}    \u{2588}\u{2588}\u{2554}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}\u{2588}\u{2588}\u{2554}\u{2550}\u{2550}\u{2588}\u{2588}\u{2557}\u{2588}\u{2588}\u{2554}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}",
    "\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}    \u{2588}\u{2588}\u{2588}\u{2554}\u{255D}     \u{2588}\u{2588}\u{2551}     \u{2588}\u{2588}\u{2551}  \u{2588}\u{2588}\u{2551}\u{2588}\u{2588}\u{2551}     ",
    "\u{2588}\u{2588}\u{2554}\u{2550}\u{2550}\u{255D}   \u{2588}\u{2588}\u{2588}\u{2554}\u{255D}      \u{2588}\u{2588}\u{2551}     \u{2588}\u{2588}\u{2551}  \u{2588}\u{2588}\u{2551}\u{2588}\u{2588}\u{2551}     ",
    "\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}    \u{255A}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2554}\u{255D}\u{255A}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2588}\u{2557}",
    "\u{255A}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}\u{255A}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}     \u{255A}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}\u{255A}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}  \u{255A}\u{2550}\u{2550}\u{2550}\u{2550}\u{2550}\u{255D}",
];

const BANNER_MIN_WIDTH: usize = 56;
const TAGLINE: &str = "Real-time CDC, radically simplified";
const LEFT_PAD: &str = "    "; // 4 spaces

/// Print the full ANSI Shadow banner with a vertical gradient.
///
/// Each line of the ASCII art gets its own color, from `PRIMARY_600` at the
/// top to `PRIMARY_400` at the bottom. Falls back to compact banner if the
/// terminal is narrower than 56 columns.
pub fn render_banner() {
    let width = terminal_width();
    if width < BANNER_MIN_WIDTH {
        render_banner_compact();
        return;
    }

    let colors = interpolate_hex(PRIMARY_600, PRIMARY_400, BANNER_LINES.len());

    println!();
    for (line, color_hex) in BANNER_LINES.iter().zip(colors.iter()) {
        let (r, g, b) = parse_hex(color_hex);
        print!("{LEFT_PAD}");
        println!("{}", style(line).color256(to_ansi256(r, g, b)).bold());
    }
    println!();

    // Tagline centered under the ASCII art.
    let art_width: usize = 44;
    let tagline_padding = LEFT_PAD.len() + (art_width.saturating_sub(TAGLINE.len())) / 2;
    let pad = " ".repeat(tagline_padding);
    let (r, g, b) = parse_hex(GRAY_400);
    println!("{pad}{}", style(TAGLINE).color256(to_ansi256(r, g, b)));
    println!();
}

/// Print a compact single-line banner: styled "ez-cdc" + version.
///
/// Used for subcommand output headers (verify, up, down, etc.).
pub fn render_banner_compact() {
    let version = env!("CARGO_PKG_VERSION");
    let (r, g, b) = parse_hex(PRIMARY_400);
    let (dr, dg, db) = parse_hex(GRAY_400);

    println!();
    println!(
        "{LEFT_PAD}{}  {}",
        style("ez-cdc").color256(to_ansi256(r, g, b)).bold(),
        style(format!("v{version}")).color256(to_ansi256(dr, dg, db)),
    );
    println!();
}

// ── Helpers ───────────────────────────────────────────────────────────────

/// Parse a "#RRGGBB" hex string into (r, g, b).
fn parse_hex(hex: &str) -> (u8, u8, u8) {
    let hex = hex.trim_start_matches('#');
    let r = u8::from_str_radix(&hex[0..2], 16).unwrap_or(0);
    let g = u8::from_str_radix(&hex[2..4], 16).unwrap_or(0);
    let b = u8::from_str_radix(&hex[4..6], 16).unwrap_or(0);
    (r, g, b)
}

/// Convert RGB to the closest ANSI 256-color code.
///
/// Uses the 6x6x6 color cube (indices 16..=231). Each channel is mapped
/// from 0..=255 to 0..=5.
fn to_ansi256(r: u8, g: u8, b: u8) -> u8 {
    let ri = (u16::from(r) * 5 / 255) as u8;
    let gi = (u16::from(g) * 5 / 255) as u8;
    let bi = (u16::from(b) * 5 / 255) as u8;
    16 + 36 * ri + 6 * gi + bi
}

/// Get terminal width, defaulting to 80 if unavailable.
fn terminal_width() -> usize {
    console::Term::stdout()
        .size_checked()
        .map(|(_, w)| w as usize)
        .unwrap_or(80)
}
